crates/typst/src/layout/flow/balance.rs

```diff
@@ -0,0 +1,215 @@
+    fn mode(balancer: &Balancer<'a, 'b, '_, Self>) -> Mode {
+        balancer.metadata
+    }
+}
+
+struct Measured;
+impl<'a, 'b> BalancerStage<'a, 'b> for Measured
+where
+    'a: 'b,
+{
+    type Data = ChildrenMetadata;
+
+    #[inline]
+    fn mode(balancer: &Balancer<'a, 'b, '_, Self>) -> Mode {
+        balancer.metadata.mode()
+    }
+}
+
+#[repr(u8, C)]
+enum ChildrenMetadata {
+    PackStart = Mode::PackStart as u8,
+    Even {
+        ideal_height: Abs,
+        children_size: HashMap<usize, Axes<Abs>>,
+    } = Mode::Even as u8,
+}
+
+impl ChildrenMetadata {
+    #[inline]
+    fn mode(&self) -> Mode {
+        unsafe {
+            // SAFETY: Docs say pointer cast of #[repr(u8)] enum to u8
+            // discriminant value is safe.
+            *(self as *const Self as *const Mode).as_ref().unwrap_unchecked()
+        }
+    }
+}
+
+pub(super) struct Balancer<'a, 'b, 'x, S>
+where
+    S: BalancerStage<'a, 'b>,
+{
+    /// All children to be balanced.
+    children: &'b [Child<'a>],
+    /// Balancing information.
+    metadata: S::Data,
+    /// Current work state.
+    work: Work<'a, 'b>,
+    /// Information about balanced columns.
+    config: &'a FlowConfig<'x>,
+    /// Index of the current column.
+    current_column: usize,
+}
+
+impl<'a, 'b, 'x, S> Balancer<'a, 'b, 'x, S>
+where
+    S: BalancerStage<'a, 'b>,
+{
+    #[inline]
+    fn mode(&self) -> Mode {
+        S::mode(self)
+    }
+}
+
+impl<'a, 'b, 'x> Balancer<'a, 'b, 'x, Initialized> {
+    pub fn new(
+        children: &'b [Child<'a>],
+        config: &'a FlowConfig<'x>,
+        mode: Mode,
+    ) -> Self {
+        Self {
+            children,
+            metadata: mode,
+            work: Work::new(&[]),
+            config,
+            current_column: 0,
+        }
+    }
+
+    fn compute_even_child_columns(
+        &self,
+        engine: &mut Engine,
+        locator: Locator,
+        bounds: Region,
+    ) -> SourceResult<ChildrenMetadata> {
+        let regions = Regions::from(bounds);
+        let mut work = Work::new(&self.children);
+        let frame = compose(engine, &mut work, self.config, locator.relayout(), regions)?;
+
+        let total_height = frame.height();
+        let ideal_height =
+            Abs::raw(total_height.to_raw() / self.config.columns.count as f64);
+
+        let mut children_size = HashMap::new();
+
+        let mut children = self.children.iter().enumerate();
+        let mut frame_items = frame.items().map(|it| it.1);
+        'outer: while let Some((i, child)) = children.next() {
+            let size = loop {
+                let item = match frame_items.next() {
+                    Some(it) => it,
+                    None => break 'outer,
+                };
+
+                // TODO: case where frame_items doesn't contain child will drain
+                // frame_items and leave all children unmeasured. This inner
+                // loop must never break to 'outer, i.e. every child must be in frame_items.
```

Comment:
Does `frame.items()` of frame returned by `compose` over some `children` always return those same `children`? If they're equal and have identical ordering, it would allow this chunk of code to be greatly simplified.